xds/internal/balancer/lrs/lrs_test.go

```diff
@@ -13,505 +13,159 @@
-		inProgress:  inprogress,
-		serverLoads: serverLoads,
+func newFakeXDSClient() *fakeXDSClient {
+	return &fakeXDSClient{
+		loadReportCh: make(chan *reportLoadArgs, 10),
 	}
 }
 
-// Equal() is needed to compare unexported fields.
-func (rcd *rpcCountDataForTest) Equal(b *rpcCountDataForTest) bool {
-	return rcd.inProgress == b.inProgress &&
-		rcd.errored == b.errored &&
-		rcd.succeeded == b.succeeded &&
-		cmp.Equal(rcd.serverLoads, b.serverLoads)
+// reportLoadArgs wraps the arguments passed to ReportLoad.
+type reportLoadArgs struct {
+	// server is the name of the server to which the load is reported.
+	server string
+	// cluster is the name of the cluster for which load is reported.
+	cluster string
+	// loadStore is the store where loads are stored.
+	loadStore interface{}
 }
 
-// equalClusterStats sorts requests and clear report internal before comparing.
-func equalClusterStats(a, b []*endpointpb.ClusterStats) bool {
-	for _, t := range [][]*endpointpb.ClusterStats{a, b} {
-		for _, s := range t {
-			sort.Slice(s.DroppedRequests, func(i, j int) bool {
-				return s.DroppedRequests[i].Category < s.DroppedRequests[j].Category
-			})
-			sort.Slice(s.UpstreamLocalityStats, func(i, j int) bool {
-				return s.UpstreamLocalityStats[i].Locality.String() < s.UpstreamLocalityStats[j].Locality.String()
-			})
-			for _, us := range s.UpstreamLocalityStats {
-				sort.Slice(us.LoadMetricStats, func(i, j int) bool {
-					return us.LoadMetricStats[i].MetricName < us.LoadMetricStats[j].MetricName
-				})
-			}
-			s.LoadReportInterval = nil
-		}
-	}
-	return cmp.Equal(a, b, cmp.Comparer(proto.Equal))
+// ReportLoad starts reporting load about clusterName to server.
+func (xdsC *fakeXDSClient) ReportLoad(server string, clusterName string, loadStore Store) (cancel func()) {
+	xdsC.loadReportCh <- &reportLoadArgs{server: server, cluster: clusterName, loadStore: loadStore}
+	return func() {}
 }
 
-func Test_lrsStore_buildStats_drops(t *testing.T) {
-	tests := []struct {
-		name  string
-		drops []map[string]uint64
-	}{
-		{
-			name: "one drop report",
-			drops: []map[string]uint64{{
-				dropCategories[0]: 31,
-				dropCategories[1]: 41,
-			}},
-		},
-		{
-			name: "two drop reports",
-			drops: []map[string]uint64{{
-				dropCategories[0]: 31,
-				dropCategories[1]: 41,
-			}, {
-				dropCategories[0]: 59,
-				dropCategories[1]: 26,
-			}},
-		},
-		{
-			name: "no empty report",
-			drops: []map[string]uint64{{
-				dropCategories[0]: 31,
-				dropCategories[1]: 41,
-			}, {
-				dropCategories[0]: 0, // This shouldn't cause an empty report for category[0].
-				dropCategories[1]: 26,
-			}},
-		},
+// waitForReportLoad waits for ReportLoad to be invoked on this client within a
+// reasonable timeout, and returns the arguments passed to it.
+func (xdsC *fakeXDSClient) waitForReportLoad() (*reportLoadArgs, error) {
+	select {
+	case <-time.After(time.Second):
+		return nil, fmt.Errorf("timeout")
+	case a := <-xdsC.loadReportCh:
```

Comment:
Why not make this `loadReportCh` of type `testutils.Channel` which provides a timed receive operation?

Reply:
Because it cannot import `testutils`..

Reply:
Hmm .... I thought it only cannot import the `fakeclient` package which imports `lrs`. But `testutils` package shouldn't have a dependency on `lrs`.

Reply:
Yes.
It is because xds-client takes lrs.Store as an input. It should be refactored to own the lrs.Store. And after that, this dependency will be gone.